File: pyFull/scripts/sources/odds_theoddsapi.py
# scripts/sources/odds_theoddsapi.py
# Fetch NBA spreads + totals from The Odds API and return in bot format.
# Requires env var: ODDS_API_KEY

import os
import re
import math
import json
import requests
import datetime
from urllib.parse import quote
from zoneinfo import ZoneInfo

_dir = os.path.dirname(os.path.abspath(__file__))
ODDS_CACHE_DIR = os.path.join(_dir, "..", "..", "..", "data", "odds_cache", "nba")

from .odds_theoddsapi_historical import fetch_closing_odds_for_game
from .espn_scoreboard import fetch_scoreboard

BASE = "https://api.the-odds-api.com/v4"


def _extract_all_espn_games(scoreboard_json):
    out = []
    for ev in (scoreboard_json or {}).get("events", []):
        comp = (ev.get("competitions") or [None])[0]
        if not comp:
            continue
        competitors = comp.get("competitors", [])
        away_c = next((c for c in competitors if c.get("homeAway") == "away"), None)
        home_c = next((c for c in competitors if c.get("homeAway") == "home"), None)
        if not away_c or not home_c:
            continue
        away = (away_c.get("team") or {}).get("displayName")
        home = (home_c.get("team") or {}).get("displayName")
        commence_time_iso = comp.get("date") or ev.get("date")
        status_name = ((comp.get("status") or {}).get("type") or {}).get("name", "")
        is_final = "FINAL" in status_name
        is_in_progress = status_name in ("STATUS_IN_PROGRESS", "STATUS_HALFTIME")
        if not away or not home:
            continue
        out.append({
            "away": away,
            "home": home,
            "commenceTimeIso": commence_time_iso,
            "isFinal": is_final,
            "isInProgress": is_in_progress,
        })
    return out


def _norm_team(name):
    return re.sub(r"\s+", " ", str(name or "")).strip()


def _today_iso_chicago():
    """Date-only in America/Chicago, formatted YYYY-MM-DD."""
    now = datetime.datetime.now(ZoneInfo("America/Chicago"))
    return now.strftime("%Y-%m-%d")


def _to_model_line(home_team, away_team, spread_points, team_for_spread):
    """
    Convert The Odds API spread into sportsbook convention (home perspective):
    -X = home favored by X, +X = away favored by X
    The Odds API already uses sportsbook convention for the named team.
    """
    if not isinstance(spread_points, (int, float)) or not math.isfinite(spread_points):
        return None

    is_home = team_for_spread == home_team
    is_away = team_for_spread == away_team

    if not is_home and not is_away:
        return None

    # If spread is for home team, return directly; if away, negate.
    if is_home:
        return spread_points
    return -spread_points


def _pick_best_bookmaker(bookmakers):
    if not isinstance(bookmakers, list) or len(bookmakers) == 0:
        return None

    preferred = ["DraftKings", "FanDuel", "BetMGM", "Caesars", "PointsBet", "BetRivers"]
    for p in preferred:
        b = next((x for x in bookmakers if x and x.get("title") == p), None)
        if b:
            return b
    return bookmakers[0]


def _find_market(bookmaker, key):
    if not bookmaker or not bookmaker.get("markets"):
        return None
    return next((m for m in bookmaker["markets"] if m and m.get("key") == key), None)


_DEFAULT_ODDS_API_KEYS = [
    "00a735b809911c5a994857dd5af3d0f2",
    "6c5699682d30fc8664737160274f8d12",
    "02a0a1d695d50185aac07fd84b965f9d",
]


def _get_odds_api_keys():
    multi = os.environ.get("ODDS_API_KEYS", "").strip()
    if multi:
        keys = [k.strip() for k in multi.split(",") if k.strip()]
        if keys:
            return keys
    single = os.environ.get("ODDS_API_KEY", "").strip()
    if single:
        return [single]
    return list(_DEFAULT_ODDS_API_KEYS)


def fetch_todays_odds():
    keys = _get_odds_api_keys()
    last_err = None
    for idx, api_key in enumerate(keys):
        url = (
            f"{BASE}/sports/basketball_nba/odds?"
            f"apiKey={quote(api_key)}"
            f"&regions=us"
            # h2h added 2026-09: the situational-systems registry has
            # moneyline systems and the feed had never carried a price for
            # them, so their entire backtest ROI was converted from the
            # spread. Costs one extra market per call.
            f"&markets=spreads,totals,h2h"
            f"&oddsFormat=american"
        )
        res = requests.get(url, timeout=30)
        if res.status_code in (401, 429):
            print(f"  [nba_odds] Key {idx+1}/{len(keys)} HTTP {res.status_code} — rotating")
            last_err = f"HTTP {res.status_code}"
            continue
        if res.status_code != 200:
            last_err = f"HTTP {res.status_code} {res.reason} {res.text[:200]}"
            continue
        data = res.json()
        if not data:
            print(f"  [nba_odds] Key {idx+1}/{len(keys)} returned empty — rotating")
            last_err = "empty response"
            continue
        break
    else:
        raise Exception(f"TheOddsAPI failed (all {len(keys)} keys): {last_err}")
    today = _today_iso_chicago()

    games = []
    now = datetime.datetime.now(datetime.timezone.utc)

    for ev in data:
        home = _norm_team(ev.get("home_team"))
        away = _norm_team(ev.get("away_team"))
        if not home or not away:
            continue

        # Filter to today (Chicago date)
        commence_str = ev.get("commence_time")
        if commence_str:
            try:
                commence = datetime.datetime.fromisoformat(commence_str.replace("Z", "+00:00"))
            except (ValueError, AttributeError):
                commence = None

            if commence:
                d_chicago = commence.astimezone(ZoneInfo("America/Chicago")).strftime("%Y-%m-%d")
                if d_chicago != today:
                    continue

                # Game already started — skip fetch, will backfill from cache
                if commence <= now:
                    continue

        book = _pick_best_bookmaker(ev.get("bookmakers"))

        line = None
        total = None

        spreads = _find_market(book, "spreads") if book else None
        totals = _find_market(book, "totals") if book else None
        h2h = _find_market(book, "h2h") if book else None

        # Prices, not just numbers. The response always carried these; the
        # parser threw them away, so every historical row grades at an assumed
        # -110 and no moneyline could be graded at all.
        away_ml = home_ml = over_ml = under_ml = None
        spread_price_away = spread_price_home = None

        if spreads and spreads.get("outcomes"):
            out = next(
                (o for o in spreads["outcomes"]
                 if isinstance(o.get("point"), (int, float)) and math.isfinite(float(o["point"]))),
                None,
            )
            if out:
                team_for_spread = _norm_team(out.get("name"))
                pts = float(out["point"])
                line = _to_model_line(home, away, pts, team_for_spread)
            for o in spreads["outcomes"]:
                nm, pr = _norm_team(o.get("name")), o.get("price")
                if not isinstance(pr, (int, float)):
                    continue
                if nm and nm == _norm_team(home):
                    spread_price_home = int(pr)
                elif nm and nm == _norm_team(away):
                    spread_price_away = int(pr)

        if totals and totals.get("outcomes"):
            out = next(
                (o for o in totals["outcomes"]
                 if isinstance(o.get("point"), (int, float)) and math.isfinite(float(o["point"]))),
                None,
            )
            if out:
                total = float(out["point"])
            for o in totals["outcomes"]:
                nm, pr = str(o.get("name") or "").lower(), o.get("price")
                if not isinstance(pr, (int, float)):
                    continue
                if nm == "over":
                    over_ml = int(pr)
                elif nm == "under":
                    under_ml = int(pr)

        if h2h and h2h.get("outcomes"):
            for o in h2h["outcomes"]:
                nm, pr = _norm_team(o.get("name")), o.get("price")
                if not isinstance(pr, (int, float)):
                    continue
                if nm and nm == _norm_team(home):
                    home_ml = int(pr)
                elif nm and nm == _norm_team(away):
                    away_ml = int(pr)

        games.append({
            "away": away,
            "home": home,
            "line": line,
            "total": total,
            "startTimeUTC": commence_str,
            "_book": book.get("title") if book else None,
            "away_ml": away_ml,
            "home_ml": home_ml,
            "over_ml": over_ml,
            "under_ml": under_ml,
            "spread_price_away": spread_price_away,
            "spread_price_home": spread_price_home,
        })

    # Load existing cache
    date_key = today.replace("-", "")
    cache_path = os.path.join(ODDS_CACHE_DIR, date_key + ".json")
    existing = {}
    try:
        if os.path.exists(cache_path):
            with open(cache_path, "r") as f:
                existing = json.load(f)
    except Exception:
        pass

    # Write fresh pre-game odds to cache
    if games:
        try:
            os.makedirs(ODDS_CACHE_DIR, exist_ok=True)
            for g in games:
                key = f"{g['away']}@{g['home']}"
                # Prices are cached alongside the numbers. Without this the
                # cache round-trip silently drops them and a game backfilled
                # from cache grades at an assumed -110 with no moneyline.
                existing[key] = {
                    "line": g["line"], "total": g["total"], "_book": g["_book"],
                    "_note": "live fetch",
                    "away_ml": g.get("away_ml"), "home_ml": g.get("home_ml"),
                    "over_ml": g.get("over_ml"), "under_ml": g.get("under_ml"),
                    "spread_price_away": g.get("spread_price_away"),
                    "spread_price_home": g.get("spread_price_home"),
                }
            with open(cache_path, "w") as f:
                json.dump(existing, f, indent=2)
            print(f"  [odds] Cached {len(games)} games to {date_key}.json")
        except Exception:
            pass

    # Backfill started/finished games from cache (so they still appear in output)
    fresh_keys = {f"{g['away']}@{g['home']}" for g in games}
    for key, val in existing.items():
        if key not in fresh_keys and val.get("line") is not None:
            parts = key.split("@", 1)
            if len(parts) == 2:
                games.append({
                    "away": parts[0], "home": parts[1], "line": val["line"],
                    "total": val.get("total"), "_book": val.get("_book"),
                    "away_ml": val.get("away_ml"), "home_ml": val.get("home_ml"),
                    "over_ml": val.get("over_ml"), "under_ml": val.get("under_ml"),
                    "spread_price_away": val.get("spread_price_away"),
                    "spread_price_home": val.get("spread_price_home"),
                })
                print(f"  [odds] Backfilled started/finished game from cache: {key}")

    return games

File: jsNBA/scripts/sources/odds_theoddsapi.mjs
// scripts/sources/odds_theoddsapi.mjs
// Fetch NBA spreads + totals from The Odds API and return in your bot format.
// Requires env var: ODDS_API_KEY

import { fetchClosingOddsForGame } from "./odds_theoddsapi_historical.mjs";
import { fetchScoreboard } from "./espn_scoreboard.mjs";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ODDS_CACHE_DIR = path.join(__dirname, "..", "..", "..", "data", "odds_cache", "nba");

const BASE = "https://api.the-odds-api.com/v4";

function extractAllESPNGames(scoreboardJson) {
  const out = [];
  for (const ev of (scoreboardJson?.events || [])) {
    const comp = ev.competitions?.[0];
    if (!comp) continue;
    const competitors = comp.competitors || [];
    const awayC = competitors.find(c => c.homeAway === "away");
    const homeC = competitors.find(c => c.homeAway === "home");
    if (!awayC || !homeC) continue;
    const away = awayC.team?.displayName;
    const home = homeC.team?.displayName;
    const commenceTimeIso = comp.date || ev.date || null;
    const statusName = comp.status?.type?.name || "";
    const isFinal = statusName.includes("FINAL");
    const isInProgress = statusName === "STATUS_IN_PROGRESS" || statusName === "STATUS_HALFTIME";
    if (!away || !home) continue;
    out.push({ away, home, commenceTimeIso, isFinal, isInProgress });
  }
  return out;
}

function normTeam(name) {
  return String(name || "").trim().replace(/\s+/g, " ");
}

function todayISOChicago() {
  // date-only in America/Chicago, formatted YYYY-MM-DD
  const fmt = new Intl.DateTimeFormat("en-CA", {
    timeZone: "America/Chicago",
    year: "numeric",
    month: "2-digit",
    day: "2-digit"
  });
  return fmt.format(new Date()); // "YYYY-MM-DD"
}

// Convert The Odds API spread into sportsbook convention:
// -X = home favored by X, +X = away favored by X
function toModelLine(homeTeam, awayTeam, spreadPoints, teamForSpread) {
  if (!Number.isFinite(spreadPoints)) return null;
  const isHome = teamForSpread === homeTeam;
  const isAway = teamForSpread === awayTeam;
  if (!isHome && !isAway) return null;
  // The Odds API already uses sportsbook convention for the named team.
  // If the spread is for the home team, return it directly.
  // If the spread is for the away team, negate to express as home line.
  if (isHome) return spreadPoints;
  return -spreadPoints;
}

function pickBestBookmaker(bookmakers) {
  if (!Array.isArray(bookmakers) || bookmakers.length === 0) return null;

  // Prefer common US books if present (otherwise first)
  const preferred = ["DraftKings", "FanDuel", "BetMGM", "Caesars", "PointsBet", "BetRivers"];
  for (const p of preferred) {
    const b = bookmakers.find((x) => x?.title === p);
    if (b) return b;
  }
  return bookmakers[0];
}

function findMarket(bookmaker, key) {
  if (!bookmaker?.markets) return null;
  return bookmaker.markets.find((m) => m?.key === key) || null;
}

export async function fetchTodaysOdds() {
  const apiKey = process.env.ODDS_API_KEY || "6c5699682d30fc8664737160274f8d12";

  // Get events (upcoming) + odds in one call
  // markets: spreads + totals
  const url =
    `${BASE}/sports/basketball_nba/odds?` +
    `apiKey=${encodeURIComponent(apiKey)}` +
    `&regions=us` +
    `&markets=spreads,totals` +
    `&oddsFormat=american`;

  const res = await fetch(url);
  if (!res.ok) {
    const txt = await res.text().catch(() => "");
    throw new Error(`TheOddsAPI failed: ${res.status} ${res.statusText} ${txt}`);
  }

  const data = await res.json();
  const today = todayISOChicago();

  const games = [];

  const now = new Date();

  for (const ev of data) {
    const home = normTeam(ev?.home_team);
    const away = normTeam(ev?.away_team);
    if (!home || !away) continue;

    // Filter to today (Chicago date)
    const commence = ev?.commence_time ? new Date(ev.commence_time) : null;
    if (commence) {
      const d = new Intl.DateTimeFormat("en-CA", {
        timeZone: "America/Chicago",
        year: "numeric",
        month: "2-digit",
        day: "2-digit"
      }).format(commence);
      if (d !== today) continue;

      // Game already started — skip fetch, will backfill from cache
      if (commence <= now) continue;
    }

    const book = pickBestBookmaker(ev?.bookmakers);

    let line = null;
    let total = null;

    const spreads = book ? findMarket(book, "spreads") : null;
    const totals = book ? findMarket(book, "totals") : null;

    if (spreads?.outcomes?.length) {
      const out = spreads.outcomes.find((o) => Number.isFinite(Number(o?.point)));
      if (out) {
        const teamForSpread = normTeam(out.name);
        const pts = Number(out.point);
        line = toModelLine(home, away, pts, teamForSpread);
      }
    }

    if (totals?.outcomes?.length) {
      const out = totals.outcomes.find((o) => Number.isFinite(Number(o?.point)));
      if (out) total = Number(out.point);
    }

    games.push({ away, home, line, total, _book: book?.title ?? null });
  }

  // Load existing cache
  const dateKey = today.replace(/-/g, "");
  const cachePath = path.join(ODDS_CACHE_DIR, dateKey + ".json");
  let existing = {};
  try {
    if (fs.existsSync(cachePath)) {
      existing = JSON.parse(fs.readFileSync(cachePath, "utf8"));
    }
  } catch {}

  // Write fresh pre-game odds to cache
  if (games.length > 0) {
    try {
      if (!fs.existsSync(ODDS_CACHE_DIR)) fs.mkdirSync(ODDS_CACHE_DIR, { recursive: true });
      for (const g of games) {
        const key = `${g.away}@${g.home}`;
        existing[key] = { line: g.line, total: g.total, _book: g._book, _note: "live fetch" };
      }
      fs.writeFileSync(cachePath, JSON.stringify(existing, null, 2));
      console.log(`  [odds] Cached ${games.length} games to ${dateKey}.json`);
    } catch {}
  }

  // Backfill started/finished games from cache
  const freshKeys = new Set(games.map(g => `${g.away}@${g.home}`));
  for (const [key, val] of Object.entries(existing)) {
    if (!freshKeys.has(key) && val.line != null) {
      const [a, h] = key.split("@");
      if (a && h) {
        games.push({ away: a, home: h, line: val.line, total: val.total, _book: val._book });
        console.log(`  [odds] Backfilled started/finished game from cache: ${key}`);
      }
    }
  }

  return games;
}